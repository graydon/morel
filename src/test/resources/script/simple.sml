(*
 * Licensed to Julian Hyde under one or more contributor license
 * agreements.  See the NOTICE file distributed with this work
 * for additional information regarding copyright ownership.
 * Julian Hyde licenses this file to you under the Apache
 * License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License.  You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.  See the License for the specific
 * language governing permissions and limitations under the
 * License.
 *)

(*) Literals
1;
~2;
~2147483647;
~2147483648;
2147483647;
"a string";
true;
[1,2,3];
(1,2);
(1,(2,"three"),true);
[(1,"a"),(2,"b"),(3,"c")];
([1,2],["a","b","c"]);

(*) Simple commands
val x = 1;
x + 2;
let
  fun fact n =
    if n = 0
    then 1
    else n * (fact (n - 1))
in
  fact 5
end;
let
  fun sum x y = x + y
in
  sum 2
end;
it 3;

(* Disabled due to bug linking recursive functions at top-level
fun str s 0 l = l
  | str s i l = str ("a" ^ s) (i - 1) (s :: l);
val str = fn : string -> int -> string list -> string list

str "" 10 [];
val it =
  ["aaaaaaaaa","aaaaaaaa","aaaaaaa","aaaaaa","aaaaa","aaaa","aaa","aa","a",""]
  : string list

(*) Note how long lists are wrapped and abbreviated
str "" 20 [];
val it =
  ["aaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaa",
   "aaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaa","aaaaaaaaaaaaaa","aaaaaaaaaaaaa",
   "aaaaaaaaaaaa","aaaaaaaaaaa","aaaaaaaaaa","aaaaaaaaa","aaaaaaaa",...]
  : string list
*)

(*) Bug: Fails due to lack of parentheses
(*
let
  fun in_ e [] = false
    | in_ e (h :: t) = e = h orelse in_ e t
in
  (in_ 3 [1,2,3], in_ 4 [1,2,3])
end;
*)

(*) Succeeds when parentheses are added
let
  fun in_ e [] = false
    | in_ e (h :: t) = e = h orelse (in_ e t)
in
  (in_ 3 [1,2,3], in_ 4 [1,2,3])
end;

(*) Multiple functions
let
  fun f x = 1
  fun g x = 2
in
  f (g 0)
end;

(*) End simple.sml
