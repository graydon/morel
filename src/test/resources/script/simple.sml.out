(*
 * Licensed to Julian Hyde under one or more contributor license
 * agreements.  See the NOTICE file distributed with this work
 * for additional information regarding copyright ownership.
 * Julian Hyde licenses this file to you under the Apache
 * License, Version 2.0 (the "License"); you may not use this
 * file except in compliance with the License.  You may obtain a
 * copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied.  See the License for the specific
 * language governing permissions and limitations under the
 * License.
 *)

(*) Literals
1;
val it = 1 : int

~2;
val it = ~2 : int

~2147483647;
val it = ~2147483647 : int

~2147483648;
val it = ~2147483648 : int

2147483647;
val it = 2147483647 : int

"a string";
val it = "a string" : string

true;
val it = true : bool

[1,2,3];
val it = [1,2,3] : int list

(1,2);
val it = (1,2) : int * int

(1,(2,"three"),true);
val it = (1,(2,"three"),true) : int * (int * string) * bool

[(1,"a"),(2,"b"),(3,"c")];
val it = [(1,"a"),(2,"b"),(3,"c")] : (int * string) list

([1,2],["a","b","c"]);
val it = ([1,2],["a","b","c"]) : int list * string list


(*) Simple commands
val x = 1;
val x = 1 : int

x + 2;
val it = 3 : int

let
  fun fact n =
    if n = 0
    then 1
    else n * (fact (n - 1))
in
  fact 5
end;
val it = 120 : int

let
  fun sum x y = x + y
in
  sum 2
end;
val it = fn : int -> int

it 3;
val it = 5 : int


(* Disabled due to bug linking recursive functions at top-level
fun str s 0 l = l
  | str s i l = str ("a" ^ s) (i - 1) (s :: l);
val str = fn : string -> int -> string list -> string list

str "" 10 [];
val it =
  ["aaaaaaaaa","aaaaaaaa","aaaaaaa","aaaaaa","aaaaa","aaaa","aaa","aa","a",""]
  : string list

(*) Note how long lists are wrapped and abbreviated
str "" 20 [];
val it =
  ["aaaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaaaa",
   "aaaaaaaaaaaaaaaa","aaaaaaaaaaaaaaa","aaaaaaaaaaaaaa","aaaaaaaaaaaaa",
   "aaaaaaaaaaaa","aaaaaaaaaaa","aaaaaaaaaa","aaaaaaaaa","aaaaaaaa",...]
  : string list
*)

(*) Bug: Fails due to lack of parentheses
(*
let
  fun in_ e [] = false
    | in_ e (h :: t) = e = h orelse in_ e t
in
  (in_ 3 [1,2,3], in_ 4 [1,2,3])
end;
*)

(*) Succeeds when parentheses are added
let
  fun in_ e [] = false
    | in_ e (h :: t) = e = h orelse (in_ e t)
in
  (in_ 3 [1,2,3], in_ 4 [1,2,3])
end;
val it = (true,false) : bool * bool


(*) Multiple functions
let
  fun f x = 1
  fun g x = 2
in
  f (g 0)
end;
val it = 1 : int
